compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala

```diff
@@ -0,0 +1,358 @@
+        } else if (isCaseApply) {
+          makeSelect(ref(defdef.symbol.owner.companionModule),
+            DefaultGetterName(termName("<init>"), n + defaultOffset))
+        } else {
+          makeSelect(This(defdef.symbol.owner.asClass),
+            DefaultGetterName(defdef.name, n + defaultOffset))
+        }
+
+        newParamSymLists
+          .take(annotatedParamListIndex)
+          .map(_.map(ref))
+          .foldLeft(inner): (lhs, newParams) =>
+            if (newParams.headOption.exists(_.isInstanceOf[TypeTree])) TypeApply(lhs, newParams)
+            else Apply(lhs, newParams)
+      )
+
+      val forwarderInner: Tree =
+        This(defdef.symbol.owner.asClass).select(defdef.symbol)
+
+      val forwarderCallArgs =
+        newParamSymLists.zipWithIndex.map{case (ps, i) =>
+          if (i == annotatedParamListIndex) ps.map(ref).take(nextParamIndex) ++ defaultCalls
+          else ps.map(ref)
+        }
+
+      val forwarderCall0 = forwarderCallArgs.foldLeft[Tree](forwarderInner){
+        case (lhs: Tree, newParams) =>
+          if (newParams.headOption.exists(_.isInstanceOf[TypeTree])) TypeApply(lhs, newParams)
+          else Apply(lhs, newParams)
+      }
+
+      val forwarderCall =
+        if (!defdef.symbol.isConstructor) forwarderCall0
+        else Block(List(forwarderCall0), Literal(Constant(())))
+
+      forwarderCall
+    }
+
+    val forwarderDef =
+      tpd.DefDef(forwarderDefSymbol, rhs = forwarderRhs())
+
+    forwarderDef.withSpan(nextSymbol.span.shift(1))
+  }
+
+  def generateFromProduct(startParamIndices: List[Int], paramCount: Int, defdef: DefDef)(using Context) = {
+    cpy.DefDef(defdef)(
+      name = defdef.name,
+      paramss = defdef.paramss,
+      tpt = defdef.tpt,
+      rhs = Match(
+        ref(defdef.paramss.head.head.asInstanceOf[ValDef].symbol).select(termName("productArity")),
+        startParamIndices.map { paramIndex =>
+          val Apply(select, args) = defdef.rhs: @unchecked
+          CaseDef(
+            Literal(Constant(paramIndex)),
+            EmptyTree,
+            Apply(
+              select,
+              args.take(paramIndex) ++
+                Range(paramIndex, paramCount).map(n =>
+                  ref(defdef.symbol.owner.companionModule)
+                    .select(DefaultGetterName(defdef.symbol.owner.primaryConstructor.name.toTermName, n))
+                )
+            )
+          )
+        } ++ Seq(
+          CaseDef(
+            Underscore(defn.IntType),
+            EmptyTree,
+            defdef.rhs
+          )
+        )
+      )
+    ).setDefTree
+  }
+
+  def generateSyntheticDefs(tree: Tree, compute: ComputeIndicies)(using Context): Option[(Symbol, Option[Symbol], Seq[DefDef])] = tree match {
+    case defdef: DefDef if defdef.paramss.nonEmpty =>
+      import dotty.tools.dotc.core.NameOps.isConstructorName
+
+      val isCaseCopy =
+        defdef.name.toString == "copy" && defdef.symbol.owner.is(CaseClass)
```

Comment:
Use `nme.copy`. Same for `"apply"` and `"fromProduct"` below.

Reply:
oops, left over from copying the code